### ISSUE TITLE
Truncation calculation doesn't consider COOKIE opt in OPT

### ISSUE THREAD

**Issue by stuartnelson3** [coredns/coredns#2581]
- CoreDNS version: 1.3.0
- Corefile:
```
. {
  prometheus :1853
  errors
  log

  loadbalance

  cache 3600 {
    prefetch 1 1m
  }

  proxy . <redacted>:53 {
    max_fails 5
    spray
  }
}
```

When performing a udp lookup against coredns, I am able to reproducibly create a truncation error. In the following commands, `+bufsize` is set to the exact size of the response.

With `+nocookie`:
```
# dig srv <srv-record> @0.0.0.0 -p1053 +ignore +bufsize=1987 +nocookie

; <<>> DiG 9.13.5 <<>> srv <srv-record> @0.0.0.0 -p1053 +ignore +bufsize=1987 +nocookie
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34981
;; flags: qr rd ra; QUERY: 1, ANSWER: 30, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1987
;; QUESTION SECTION:
;<srv-record>. IN SRV

;; ANSWER SECTION:
// elided ...

;; Query time: 16 msec
;; SERVER: 127.0.0.1#1053(0.0.0.0)
;; WHEN: Tue Feb 19 18:05:37 CET 2019
;; MSG SIZE  rcvd: 1987
```
The response perfectly fits in the buffer.

When I execute the same command with default behavior (`+cookie`):
```
# dig srv <srv-record> @0.0.0.0 -p1053 +ignore +bufsize=1987

; <<>> DiG 9.13.5 <<>> srv <srv-record> @0.0.0.0 -p1053 +ignore +bufsize=1987
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33357
;; flags: qr rd ra; QUERY: 1, ANSWER: 30, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1987
; COOKIE: 535328c17aaaf658 (echoed)
;; QUESTION SECTION:
;<srv-record>. IN SRV

;; ANSWER SECTION:
// elided ...

;; Query time: 19 msec
;; SERVER: 127.0.0.1#1053(0.0.0.0)
;; WHEN: Tue Feb 19 18:11:28 CET 2019
;; MSG SIZE  rcvd: 1999
```

The message is not truncated, and exceeds the requested buffer size.

This appears to be an issue in the `Scrub()` method. When checking for `EDNS0`, it assumes that there is only a single opt record.

https://github.com/coredns/coredns/blob/master/request/request.go#L273-L277

**Comment by stuartnelson3** [comment]
After some investigation, the problem appears to be that the compressed length as calculated here:
https://github.com/coredns/coredns/blob/master/request/request.go#L266
results in a value that is less than what is actually written to the client.

When the request isn't cached:

- requested size 1987
- uncompressed size 3096
- compressed size 1936

When the request is cached:

- requested size 1987
- uncompressed size 3176
- compressed size 1976

When the request is not cached, 29/30 records are returned. This is correctly truncated with or without `+cookie`.

When the request is cached, 30/30 records are returned. This perfectly fills the requested buffer with `+nocookie`, and exceeds the buffer with `+cookie`.

Looking at `miekg/dns`, it appears the `msgLenWithCompressionMap` (https://github.com/miekg/dns/blob/master/msg.go#L1006) has no `dns.Extra` records when calculated with a cached record. Why, I'm not sure.

**Comment by miekg** [comment]
Amended the title, because additional OPT records are not allowed, you can only have one

**Comment by stuartnelson3** [comment]
:+1: thanks for fixing that

**Comment by miekg** [comment]
yes, this is a true issue the opt record doesn't have a fixed size (we've got some `const` in the code right now)

**Comment by miekg** [comment]
this should be fixed by fixing #2728
